### DISCUSSION TITLE
adaptable equity size

### DISCUSSION THREAD

**Discussion by Hawar-Dzaee** [kernc/backtesting.py#572] [Q&A]
if you set self.buy( size= 1000), it will use 1000 on each trade, how can you use 1 percent of your current equity for each trade?
what I am asking to use (1000*0.01), and let say u it hit your SL and your equity became 990 since ( you only risk 1 percent of 1000, which means you risked 10$. now on the second trade, I want to risk 9.90 which is 1 percent of my current equity, and so on. 
how do I do that?

**Comment by kernc** [comment]
From [`Order.size` docs](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Order.size):
> If size is a value between 0 and 1, it is interpreted as a fraction of current available liquidity

So:
```py
# Buy for 1% of equity
self.buy(size=0.01)

**Reply by Hawar-Dzaee** [reply]
self.buy(size=0.01) 
cash=10000
margin argument has not been set.
commission =0.00
shouldn't it risk 100$ per trade?  and from there it changes 
yet I got this. 
also,

I noticed, why (0 index trade) comes before (1 index trade) shouldn't be vice versa?
notice the entry time of each.
I really appreciate clarification 

![Screenshot_5](https://user-images.githubusercontent.com/96496172/149541234-57acbb5e-256b-41b6-ba13-216d1fc50d45.png)

**Reply by kernc** [reply]
> shouldn't it risk 100$ per trade?

80 × $1.22518 is roughly $100 for the first trade, no?



> I noticed, why (0 index trade) comes before (1 index trade) shouldn't be vice versa?
> notice the entry time of each.

Can't say for sure, but trades in the result data frame are ordered by the time of closing:
https://github.com/kernc/backtesting.py/blob/94d20da85e278102a0cc71b27a3a35b815e11648/backtesting/backtesting.py#L957
I say your later order hit its SL and was closed earlier than the first one.

**Reply by Hawar-Dzaee** [reply]
so does that mean I lost $100 on my first trade? if not how can do that? 
my tp hit when my profit reaches 1 percent of my current liquidity( that is gaining $100 in this case) and sl hits when it reaches 1 percent loss ( in our case -100$  on current liquidity), and I don't mind how much the entry size is, as long as when it hits my tp I am gaining 1 percent on my current liquidity and losing 1 percent when it hit my sl, I know I can be a bit confusing but is there a way to do that?
I did set specific tp and sl but I want to set that kind of position sizing. 
I went through the documentation many times, couldn't figure it out.
this is just a random example.
![Screenshot_1](https://user-images.githubusercontent.com/96496172/149549429-819dafd4-d597-42a2-ae72-89b17193ed81.png)
please if my question isn't clear enough ask for more details, and as always thank you for your time.

**Reply by kernc** [reply]
You lost $0.13840 on the first trade as it says in the PnL column. In order for 1% SL (or TP) to represent exactly 1% of current equity, just leave `size=` out, e.g.:
```py
self.buy(sl=0.99*price, tp=1.01*price)
```

**Reply by Hawar-Dzaee** [reply]
This is the equation 
pips*lots*10 = equity*0.01 (that is to be one percent of the equity)
I want to find out the number of lots to use that if it hits my sl, I lose 1 percent of my current capital.
in this case, its 40 pips and my capital is $10,000 meaning, so the calculation is as follows:-
40(pips)*lots(which is the variable)* 10 = 10000(my capital)* 0.01(the percentage of my capital that I want to risk)
lots = 0.25 lots ( so if i take a position sizing of 0.25 
![Screenshot_1](https://user-images.githubusercontent.com/96496172/149618796-07ce3acc-bd08-4dbb-a119-7bb5f82e86bb.png)

and if hits my tp or sl i would gain or loss $100 dollars of my capital, so for the first row of the PnL column should be (either -100 or +100)
and so on the following trades.
![Screenshot_2](https://user-images.githubusercontent.com/96496172/149618835-9aee4d4d-36ef-4cba-aa63-89df53b040c0.png)

I have already specified conditions for tp and sl 
![Screenshot_3](https://user-images.githubusercontent.com/96496172/149618915-ad10e2f8-d532-4ad5-87d8-9e45f3ad1792.png)

also comes to my mind, how to fix a price after entering a position. because if your tp **talib.MAX(x, timeperiod=30)**  is going to change, how to if I fix this number at the time of entering a position. so if that **talib.MAX(x, timeperiod=30)** is 1.2345 at the time of entering a position I want to be my tp and not to be changeable until it hits my tp 

I've been working on this problem for weeks now :( , I would be very grateful for anykind of help :)

**Reply by kernc** [reply]
So what you're looking for is:

    (entry_price - sl_price) * size == 0.01 * self.equity

Thereafter, it's a simple case of working out the equation to solve for `size`. See also [`Strategy.equity`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Strategy.equity).

-------

[`Trade` properties `.sl` and `.tp`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Trade.tp) are _writable_. They are set (to a fixed value) when placing the order, but you can update them on a trade at a later time if you wish.

**Reply by Hawar-Dzaee** [reply]
seems right, but how do I implement this, and where should do I pass this argument
like this self.buy((entry_price - sl_price) * size == 0.01 * self.equity) ? under next? 
are they built-in variables or I should define them?
and if they are built-in and which library should I import, because it's throwing me an undefined error.
